### PR TITLE
"[QUIC] Adopted msquic generated interop (#68288)" attempt no 2

### LINKED ISSUE #67377: [QUIC] Adopt msquic generated interop layer
Splitting work from https://github.com/dotnet/runtime/issues/64554 since it's rather non-trivial change and should be solved separately.

### PR DESCRIPTION
**Pull request by ManickaP**
This reverts commit 4820674a4471872d742c380f38e766b80dc79385.

Brings back #68288

Fixes https://github.com/dotnet/runtime/issues/67377, https://github.com/dotnet/runtime/issues/67301, https://github.com/dotnet/runtime/issues/68954

### CODE REVIEW

**Comment by ManickaP** [comment]
So the AOT itself is not failing anymore, but now I get:
```
Failed to load AOT module '/root/helix/work/correlation/System.Net.Quic.dll.so' ('/root/helix/work/correlation/System.Net.Quic.dll') in aot-only mode.
```
In Work Item JIT.Regression.JitBlue: https://helixre8s23ayyeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-69009-merge-bd97177d9b0441c6b0/JIT.Regression.JitBlue/1/console.ca962c5e.log?helixlogtype=result

@akoeplinger could help with this or do you know who could?

**Comment by ManickaP** [comment]
I have a fix for #68954 ready, but I don't want to stop the pipeline by pushing it right now.

File: src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs [hunk 1]
```diff
@@ -143,19 +145,37 @@ private async Task HandleControlStreamAsync(Http3LoopbackStream controlStream)
...
 
-                // Must be the control stream
+                // Must be the control stream.
                 await HandleControlStreamAsync(stream);
+
+                // We've already accepted request stream --> return it.
+                if (requestStream is not null)
+                {
+                    return requestStream;
+                }
```
Comment (ManickaP) [inline src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs:178]: This is fix for the assert problem from https://github.com/dotnet/runtime/issues/68954

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs [hunk 2]
```diff
@@ -903,38 +897,40 @@ private void Dispose(bool disposing)
                     ExceptionDispatchInfo.SetCurrentStackTrace(new QuicOperationAbortedException("Read was canceled")));
             }
 
+
+            // Check if we already got final event.
+            bool releaseHandles = Interlocked.Exchange(ref _state.ShutdownDone, State.ShutdownDone_Disposed) == State.ShutdownDone_NotificationReceived;
             if (releaseHandles)
             {
                 _state.Cleanup();
             }
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs:907]: This together with lines 871-877 in the original are a fix for the use after free part in https://github.com/dotnet/runtime/issues/68954
Note that the removed: `_state.ShutdownState = ShutdownState.Finished;` is actually set in `_state.Cleanup();`:
https://github.com/dotnet/runtime/blob/6ca8c9bc0c4a5fc1082c690b6768ab3be8761b11/src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs#L94

File: src/tests/build.proj [hunk 3]
```diff
@@ -124,7 +124,7 @@
     </RemoveDuplicates>
     <ItemGroup>
       <TestsAndAssociatedAssemblies Include="%(TestDirs.Identity)/*.dll" Exclude="@(NoMonoAotAssemblyPaths)" />
-      <CoreRootDlls Include="$(CORE_ROOT)/*.dll" Exclude="$(CORE_ROOT)/xunit.performance.api.dll" />
+      <CoreRootDlls Include="$(CORE_ROOT)/*.dll" Exclude="$(CORE_ROOT)/xunit.performance.api.dll;$(CORE_ROOT)/System.Net.Quic.dll" />
```
Comment (ManickaP) [inline src/tests/build.proj:127]: Excludes us from mono AOT

File: src/tests/issues.targets [hunk 4]
```diff
@@ -3026,6 +3026,19 @@
...
+        </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/JIT/Regression/JitBlue/GitHub_25468/GitHub_25468/**">
+            <Issue>https://github.com/dotnet/runtime/issues/52977</Issue>
+        </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/JIT/CheckProjects/CheckProjects/**">
+            <Issue>https://github.com/dotnet/runtime/issues/52977</Issue>
+        </ExcludeList>
+        <ExcludeList Include="$(XunitTestBinBase)/ JIT/Performance/CodeQuality/BenchmarksGame/k-nucleotide/k-nucleotide-9/**">
+            <Issue>https://github.com/dotnet/runtime/issues/52977</Issue>
+        </ExcludeList>
```
Comment (ManickaP) [inline src/tests/issues.targets:3041]: Skips tests failing due to missing mono AOT generated System.Net.Quic.dll.so.

**Comment by ManickaP** [comment]
The pipeline is green and mono llvmfullaot ran: https://dev.azure.com/dnceng/public/_build/results?buildId=1763129&view=results

Can I get a review here? The PR is same as https://github.com/dotnet/runtime/pull/68288 with the exceptions marked with comments, which are fixed for mono AOT error and #68954

**Review by rzikm — APPROVED** [review]
LGTM!

**Pull request merged**